ArduPlane/pullup.cpp

```diff
@@ -0,0 +1,244 @@
+                            ahrs.pitch_sensor*0.01,
+                            ahrs.roll_sensor*0.01,
+                            current_loc.alt*0.01);
+            stage = Stage::WAIT_LEVEL;
+        }
+        return false;
+    }
+
+    case Stage::PUSH_NOSE_DOWN: {
+        if (fabsf(ahrs.roll_sensor*0.01) < aparm.roll_limit) {
+            stage = Stage::WAIT_LEVEL;
+        }
+        return false;
+    }
+
+    case Stage::WAIT_LEVEL: {
+        // When pitch has raised past lower limit used by speed controller, wait for airspeed to approach
+        // target value before handing over control of pitch demand to speed controller
+        bool pitchup_complete = ahrs.pitch_sensor*0.01 > MIN(0, aparm.pitch_limit_min);
+        const float pitch_lag_time = 1.0f * sqrtf(ahrs.get_EAS2TAS());
+        float aspeed;
+        const float aspeed_derivative = (ahrs.get_accel().x + GRAVITY_MSS * ahrs.get_DCM_rotation_body_to_ned().c.x) / ahrs.get_EAS2TAS();
+        bool airspeed_low = ahrs.airspeed_estimate(aspeed) ? (aspeed + aspeed_derivative * pitch_lag_time) < 0.01f * (float)plane.target_airspeed_cm : true;
+        bool roll_control_lost = fabsf(ahrs.roll_sensor*0.01) > aparm.roll_limit;
+        if (pitchup_complete && airspeed_low && !roll_control_lost) {
+                gcs().send_text(MAV_SEVERITY_INFO, "Pullup level r=%.1f p=%.1f alt %.1fm AMSL",
+                                ahrs.roll_sensor*0.01, ahrs.pitch_sensor*0.01, current_loc.alt*0.01);
+                break;
+        } else if (pitchup_complete && roll_control_lost) {
+                // push nose down and wait to get roll control back
+                gcs().send_text(MAV_SEVERITY_ALERT, "Pullup level roll bad r=%.1f p=%.1f",
+                                ahrs.roll_sensor*0.01,
+                                ahrs.pitch_sensor*0.01);
+                stage = Stage::PUSH_NOSE_DOWN;
+        }
+        return false;
+    }
+    case Stage::NONE:
+        break;
+    }
+
+    // all done
+    stage = Stage::NONE;
+    return true;
+}
+
+/*
+  stabilize during pullup from balloon drop
+ */
+void GliderPullup::stabilize_pullup(void)
+{
+    const float speed_scaler = plane.get_speed_scaler();
+    switch (stage) {
+    case Stage::WAIT_AIRSPEED: {
+        SRV_Channels::set_output_scaled(SRV_Channel::k_elevator, elev_offset*4500);
+        SRV_Channels::set_output_scaled(SRV_Channel::k_rudder, 0);
+        plane.nav_pitch_cd = 0;
+        plane.nav_roll_cd = 0;
+        uint32_t now = AP_HAL::millis();
+        if (now - plane.last_stabilize_ms > 1000) {
+            plane.pitchController.reset_I();
+            plane.yawController.reset_I();
+        }
+        plane.last_stabilize_ms = now;
+        SRV_Channels::set_output_scaled(SRV_Channel::k_aileron, plane.rollController.get_rate_out(0, speed_scaler));
+        plane.logger.Write_PID(LOG_PIDR_MSG, plane.rollController.get_pid_info());
```

Comment:
I don't think we should just start logging at loop rate. Even if we should this is a odd way of doing it because the existing PID logging is still going on, so there will be lots of duplication.

Reply:
removed, this was primarily for debugging